### PR TITLE
feat: add fine-tuning validation frequency

### PR DESCRIPTION
**Pull request by anuragg1209**
## Issue
Please link the corresponding GitHub issue. If an issue does not already exist,
please open one to describe the bug or feature request before creating a pull request.

This allows us to discuss the proposal and helps avoid unnecessary work.

## Motivation and Context

This PR allows users to reduce validation overhead while preserving checkpoint recovery and early-stopping behavior at scheduled validation checks. 

## Public API Changes

-   [x] No Public API changes
-   [ ] Yes, Public API changes (Details below)

---

## How Has This Been Tested?

Tested locally and also added a test case. 

## Checklist

-   [x] The changes have been tested locally.
-   [x] Documentation has been updated (if the public API or usage changes).
-   [x] A changelog entry has been added (see `changelog/README.md`), or "no changelog needed" label requested.
-   [x] The code follows the project's style guidelines.
-   [x] I have considered the impact of these changes on the public API.

---

### CODE REVIEW

**Comment by anuragg1209** [comment]
This PR addresses issue #811 and PR #1098.

**Comment by anuragg1209** [comment]
Hi @bejaeger, here’s a small PR adding a validation_frequency option for fine-tuning. I’ve kept the scope focused on scheduling validation while preserving the existing default behavior (validation_frequency=1).

A couple of design points I’d appreciate your input on:

1. If early_stopping=True, validation runs every N epochs, and time_limit stops training before the first such validation, training has updated the weights but early stopping has not evaluated them. The current code then restores the original pre-training weights at the end, discarding those updates. This is an edge case; I have not addressed it in this PR to keep the scope small.

2. With validation_frequency > 1, I define `early_stopping_patience` in units of validation checks rather than training epochs. Skipped epochs do not consume patience, so early stopping decisions are based only on observed validation metrics. This felt like the most consistent interpretation, but I’d welcome your thoughts.

File: src/tabpfn/finetuning/finetuned_base.py [hunk 1]
```diff
@@ -1160,7 +1207,7 @@ def _ddp_broadcast_primary_metric(metric: float) -> float:
                     patience_counter += 1
                     if is_main_process:
                         logger.info(
-                            "⚠️  No improvement for %s epochs. Best %s: %.4f",
+                            "⚠️  No improvement for %s validation checks. Best %s: %.4f",
```
[the remark this answers is not in the document]
Comment (anuragg1209) [reply src/tabpfn/finetuning/finetuned_base.py:1186]: This is the same point that I have raised in the comment above. To be discussed with @bejaeger.

[the remark this answers is not in the document]
Comment (bejaeger) [reply src/tabpfn/finetuning/finetuned_base.py:1186]: This sounds like a usage "error" to me and I think it's fine not to do anything custom in this case.

File: src/tabpfn/finetuning/finetuned_base.py [hunk 2]
```diff
@@ -603,6 +612,23 @@ def _get_train_val_split(
...
+        """Return validation_frequency as a positive integer."""
+        if isinstance(self.validation_frequency, bool):
+            raise ValueError("`validation_frequency` must be a positive integer.")
+
+        try:
+            validation_frequency = operator.index(self.validation_frequency)
+        except TypeError as error:
+            raise ValueError(
+                "`validation_frequency` must be a positive integer."
+            ) from error
```
Comment (bejaeger) [inline src/tabpfn/finetuning/finetuned_base.py]: why don't we rely on type hints here? seems we add unnecessary complexity here.

File: src/tabpfn/finetuning/finetuned_base.py [hunk 3]
```diff
@@ -1099,11 +1141,13 @@ def _ddp_broadcast_primary_metric(metric: float) -> float:
                 _logger.log_epoch(epoch_log_metrics, step=global_step)
 
                 primary_metric = eval_result.primary
-            else:
+            elif run_validation:
```
Comment (bejaeger) [inline src/tabpfn/finetuning/finetuned_base.py]: doesn't this lead to doing the evaluation on all the ranks? why did you introduce this?

File: src/tabpfn/finetuning/finetuned_base.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (anuragg1209) [reply src/tabpfn/finetuning/finetuned_base.py:1186]: Sounds good. Marking this as resolved.

File: src/tabpfn/finetuning/finetuned_base.py (hunk 3 above)
Comment (anuragg1209) [reply src/tabpfn/finetuning/finetuned_base.py]: Fixed.

**Review by bejaeger — APPROVED** [review]
Sorry for the late review, LGTM!
